submarineswap.go

```diff
@@ -0,0 +1,265 @@
+		}
+
+		var key bytes.Buffer
+		_, err = key.WriteString("swapper:")
+		if err != nil {
+			return err
+		}
+		_, err = key.Write(hash)
+		if err != nil {
+			return err
+		}
+
+		value := bucket.Get(key.Bytes())
+		if value == nil {
+			return errors.New("Not found")
+		}
+
+		submarineData := bytes.NewBuffer(value)
+		savedNetID, err := submarineData.ReadByte()
+		if err != nil {
+			return err
+		}
+		if savedNetID != netID {
+			return errors.New("Not the same network")
+		}
+		b := make([]byte, 16)
+		_, err = submarineData.Read(b)
+		if err != nil {
+			return err
+		}
+		creationHeight = int64(binary.BigEndian.Uint64(b[0:]))
+		lockHeight = int64(binary.BigEndian.Uint64(b[8:]))
+
+		swapperKey = make([]byte, btcec.PrivKeyBytesLen)
+		_, err = submarineData.Read(swapperKey)
+		if err != nil {
+			return err
+		}
+
+		script = make([]byte, submarineData.Len())
+		_, err = submarineData.Read(script)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}, func() {})
+
+	return
+}
+
+func NewSubmarineSwap(wdb walletdb.DB, manager *waddrmgr.Manager, net *chaincfg.Params,
+	chainClient chain.Interface, c *channeldb.ChannelStateDB, pubKey, hash []byte) (address btcutil.Address, script, swapperPubKey []byte, lockHeight int64, err error) {
+
+	if len(pubKey) != btcec.PubKeyBytesLenCompressed {
+		err = errors.New("pubKey not valid")
+		return
+	}
+
+	if len(hash) != 32 {
+		err = errors.New("hash not valid")
+		return
+	}
+
+	//Need to check that the hash doesn't already exists in our db
+	_, _, _, _, errGet := getSwapperSubmarineData(c, net.ScriptHashAddrID, hash)
+	if errGet == nil {
+		err = errors.New("Hash already exists")
+		return
+	}
+
+	//Create swapperKey and swapperPubKey
+	key, err := btcec.NewPrivateKey(btcec.S256())
+	if err != nil {
+		return
+	}
+	swapperKey := key.Serialize()
+	swapperPubKey = key.PubKey().SerializeCompressed()
+	lockHeight = defaultLockHeight
+
+	//Create the script
+	script, err = genSubmarineSwapScript(swapperPubKey, pubKey, hash, defaultLockHeight)
+	if err != nil {
+		return
+	}
+
+	currentHash, currentHeight, err := chainClient.GetBestBlock()
+	if err != nil {
+		return
+	}
+
+	address, err = importScript(
+		wdb,
+		manager,
+		net,
+		currentHeight,
+		*currentHash,
+		script,
+	)
+	if err != nil {
+		return
+	}
+	//Watch the new address
+	err = chainClient.NotifyReceived([]btcutil.Address{address})
+	if err != nil {
+		return
+	}
+	//Need to save the data keyed by hash
+	err = saveSwapperSubmarineData(c, net.ScriptHashAddrID, hash, int64(currentHeight), lockHeight, swapperKey, script)
+
+	return
+}
+func newAddressWitnessScriptHash(script []byte, net *chaincfg.Params) (*btcutil.AddressWitnessScriptHash, error) {
+	witnessProg := sha256.Sum256(script)
+	return btcutil.NewAddressWitnessScriptHash(witnessProg[:], net)
+}
+func importScript(db walletdb.DB, manager *waddrmgr.Manager, net *chaincfg.Params, startHeight int32, startHash chainhash.Hash, script []byte) (btcutil.Address, error) {
```

Comment:
Here we need to clean all lnd/walletdb dependency.
We need to extract out only the code that converts the script to address. We should also rename the function accordingly (createAddressFromScript)